### ISSUE TITLE
Consider enabling semantics in tests and debug mode always

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#34378]
See https://github.com/flutter/flutter/issues/31139

We have asserts in various places (and want to add more) to guard against incorrect semantics constructions.  Violating these asserts can result in crashes in the a11y bridge implementations at runtime in release mode.

However, even adding more asserts could still result in developers not seeing the assertion errors in their apps unless they test in an environment where accessibility services are enabled - e.g. the iOS Simulator where we always turn them on, or an Android device with accessibility settings enabled.  Unfortunately, it is very easy to test only on an Android device and find that everything appear to work when it does not.

We should consider enabling semantics for all tests and in debug mode, so that developers can see these errors firing and be more likely to fix or at least report them before going to production.

/cc @goderbauer @jonahwilliams @Hixie @tvolkert

**Comment by goderbauer** [comment]
I like the idea. Obviously, just like with any crash, we have to make sure its obvious why people a re seeing a crash.

**Comment by dnfield** [comment]
I'm wondering if we should give developers the option to turn this off - though I suspect it would be better to just make sure that hot restart/hot reload play nicely with clearing/refreshing the semantics tree.  I've fixed bugs related to that but there may be more.

**Comment by goderbauer** [comment]
Yeah, everything should "just work" :)

We should probably also include this change in the changelog and maybe tell people on flutter-announce. so nobody is surprised if they see more asserts.

**Comment by jonahwilliams** [comment]
We probably don't want this on by default for the web though, until we make performance improvements

**Comment by dnfield** [comment]
The change I'm proposing in Flutter/engine is completely in Native code - this will have no impact on flutter for web.

**Comment by jonahwilliams** [comment]
Right, but we also probably want to consider consistency of experience

**Comment by jonahwilliams** [comment]
I guess that is already not consistent because of how the iOS simulator behaves, so it probably doesn't matter

**Comment by dnfield** [comment]
Testing is done, at least for widget tests.

I'm running into some challenges with trying to do this for actual apps.

It's easy enough to get the framework to generate the tree. It's not much harder to get the engine to tell the framework to do that.

The problem right now is that each embedding listens to the OS for whether accessibility services are enabled, and if they're not the embedding doesn't listen to accessibility events. So while turning this on for apps would exercise the framework side semantics code, it would not exercise the bridge implementation for the platform, which is the kind of bug i was hoping to get to.

Another way to approach this is to write tests for the variants of the public API of the bridges. It would probably be better to do that for now, though I think this would be worth coming back to if we can figure out a way to make it worthwhile (or if we're convinced that at least turning on the framework side semantics code is worthwhile).
